### DISCUSSION TITLE
Features for protein design

### DISCUSSION THREAD

**Discussion by dacarlin** [biotite-dev/biotite#483] [Ideas]
Biotite is a popular choice in the protein design and protein ML community. For example, [biotite is used extensively in training and inference for Meta's evolutionary-scale transformer models (ESM) protein transformer models](https://github.com/search?q=repo%3Afacebookresearch%2Fesm%20biotite&type=code). Are there features we could add to biotite to expand its functionality to protein design workflows? 

A few ideas might be: 

- simple functions for mutating/changing sequences based on biochemical nomenclature 
- tools for creating synthetic genes for a taxon 
- interfaces for popular deep learning tools such as ESM 

I would be interested to hear if others have thoughts about potentially adding some of these functions to biotite!

**Comment by padix-key** [comment]
Currently `biotite.structure`/`biotite.sequence` covers functionality for analyzing, editing and (to a limited extent) visualize structures/sequences. Although I think going also into the modelling direction sounds intriguing to me, it would also open a new large topic and therefore broaden the scope too much, in my opinion. However, I think such functionalities can still find a place in the Biotite ecosystem:

1. If the functionality is not specific to modelling/ML, but has a broader area of application in structure/sequence analysis/editing, it would be a welcome addition to `biotite.structure`/`biotite.sequence`.
2. Otherwise, such functionality can be published as example in the gallery. Although, this makes the functionality not as readily usable as a function in Biotite itself, some examples are still quite often adapted in actual research (e.g. [this one](https://www.biotite-python.org/examples/gallery/structure/transketolase_sse.html)).
3. Larger sets of functionalities acting on Biotite's data structures can also be released as [extension packages](https://www.biotite-python.org/extensions.html) and linked in the Biotite documentation.
4. If the functionality is an interface to a popular tool or database, such as ESM or its Metagenomic atlas, its fits well into the `biotite.application` or `biotite.database` subpackages.

On the first glance I would think that

> - simple functions for mutating/changing sequences based on biochemical nomenclature

and

> - tools for creating synthetic genes for a taxon

would fit into category 2, while

> - interfaces for popular deep learning tools such as ESM

would fit into category 4.

I am really interested into discuss with you, which specfic features you have in mind :smiley:.

**Reply by dacarlin** [reply]
This is great, thanks for the detailed explanation of your thinking! 

It seems like accessing the structures from the ESM Metagenomic Atlas is very similar to adding support for the AlphaFold database (category 4), agree. Perhaps we can start a mini-project to add support for fetching ESM Metagenomic Atlas structures along the same lines as the mini-project for adding AlphaFold that we are currently working on together. 

I like the idea of the category 2 ideas as examples! I can definitely cook some up for making mutations to sequences and creating native-like nucleotide encodings.

More broadly, there is a lot of overlap between tasks in bioinformatics and tasks in protein design. Here’s an example. If you use a protein design algorithm like ProteinMPNN on your structure, it will output a FASTA file that is an alignment containing the thousands of designed/sampled sequences. A key piece of using ProteinMPNN is actually reading in that alignment, calculating statistics, then labeling the sequences and subsampling. These are all things that biotite already supports, which is part of why it’s so useful already for protein design. So that got me thinking, are there other small additions we can make to biotite to round out that toolset? 

However, since biotite is primarily a tool analyzing existing sequences (not generating new ones), I agree that it would majorly change the scope of the project to go fully into the modeling/design direction.

**Comment by padix-key** [comment]
> It seems like accessing the structures from the ESM Metagenomic Atlas is very similar to adding support for the AlphaFold database (category 4), agree. Perhaps we can start a mini-project to add support for fetching ESM Metagenomic Atlas structures along the same lines as the mini-project for adding AlphaFold that we are currently working on together.

If you like to contribute the interface, I would be happy to review :smiley:. I am not sure, but I think that Metagenomic Atlas provides two functionalities, downloading of already existing structures and de novo folding of submitted sequences, right? In my opinion the former could go into `biotite.database` because it is just fast data retrieval from database, while the latter could go into `biotite.application` as it really calculates a result from input (like e.g. NCBI BLAST implemented in `biotite.application.blast.BlastWebApp`).

> I like the idea of the category 2 ideas as examples! I can definitely cook some up for making mutations to sequences and creating native-like nucleotide encodings.

Sounds good!

> So that got me thinking, are there other small additions we can make to biotite to round out that toolset?

If these are general purpose functions, they could fit well. We could discuss that on a case-by-case basis. What I think goes to far, are functions that are directly involved in modelling, for example predicting structures or calculating energies.
